### PR TITLE
bpf: Support proxy upstream connections with original source address/port

### PR DESCRIPTION
**Pull request by jrajahalme**
Allow Envoy egress proxies to use the original source address and port in the upstream connections. This allows destination node to map the source IP to the policy ID. Egress Kafka is not supported yet, as Kafka only works for ingress at the moment. DNS proxy (always egress) support for original source address is left for a follow-up PR.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8870)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/datapath/iptables/iptables.go [hunk 1]
```diff
@@ -552,7 +552,41 @@ func (m *IptablesManager) InstallRules(ifName string) error {
...
+		}
+	}
+
 	if option.Config.EnableIPv4 {
+		// See comment above for the IPv6 case.
+		if err := runProg("iptables", append(
+			m.waitArgs,
+			"-t", "mangle",
+			"-A", ciliumPreMangleChain,
+			"-m", "socket", "--transparent", "--nowildcard",
```
Comment (joestringer) [inline pkg/datapath/iptables/iptables.go:653]: (Note, not a question): These rules overlap with the same ones being added in https://github.com/cilium/cilium/pull/8864.

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -460,6 +466,12 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 	reason = ret;
 
+	// Check it this is return traffic to an ingress proxy.
```
Comment (joestringer) [inline bpf/bpf_lxc.c:469]: The comment says "ingress proxy" but the code below is setting the mark for egress proxy too..?[0]

File: bpf/bpf_lxc.c [hunk 3]
```diff
@@ -500,7 +512,7 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 		 */
 		ct_state_new.src_sec_id = SECLABEL;
 		ret = ct_create4(get_ct_map4(&tuple), &tuple, skb, CT_EGRESS,
-				 &ct_state_new);
+				 &ct_state_new, verdict > 0);
```
Comment (joestringer) [inline bpf/bpf_lxc.c:515]: [0] This is in handling packets "from_lxc", ie from the endpoint, so the verdict here (if nonzero) is an egress L7 policy.

File: bpf/bpf_lxc.c [hunk 4]
```diff
@@ -790,6 +802,14 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, s
 
 	*reason = ret;
 
+	// Check it this is return traffic to an egress proxy.
+	// Do not redirect again if the packet is coming from the egress proxy.
+	if ((ret == CT_REPLY || ret == CT_RELATED) && ct_state.proxy_redirect &&
+	    !tc_index_skip_egress_proxy(skb)) {
```
Comment (joestringer) [inline bpf/bpf_lxc.c:808]: I'm confused. This function is `ipv6_policy()` which is called from `tail_ipv6_policy()` and `tail_ipv6_to_endpoint()`, ie the endpoint ingress case. Why is an egress proxy skip bit being checked here in the ingress path?

**Review by joestringer — CHANGES_REQUESTED** [review]
I took a quick look over the changes, and I'm a bit surprised by some of them, questions below.

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (jrajahalme) [reply bpf/bpf_lxc.c:469]: reply direction traffic from an LXC, if going to a local proxy, is for the ingress proxy. Original direction traffic from an LXC, if going to a local proxy, if for the egress proxy :-)

File: bpf/bpf_lxc.c (hunk 4 above)
Comment (jrajahalme) [reply bpf/bpf_lxc.c:808]: See above.

File: bpf/bpf_lxc.c (hunk 3 above)
Comment (jrajahalme) [reply bpf/bpf_lxc.c:515]: See above

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/datapath/iptables/iptables.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/datapath/iptables/iptables.go:653]: Right, will rebase on top of #8864 when it merges.

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
Works like a charm in my cluster with egress + ingress policy.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Using the original source address for the dns proxy may be a stretch. Basic problem that two consecutive requests from the same source address:port will cause two client sockets to be opened, now with the same address:port, and the second one failed with `bind: address already in use`. I'm testing if SO_REUSEADDR makes any difference, but still `miekg/dns` expects the response to arrive on the same socket as the request was sent on, so even if two sockets could bind to the same address, it likely does not work.

A proper solution would use one socket for each source address:port, and then multiplex multiple queries on top of it.

**Comment by jrajahalme** [comment]
It may be better to split the DNS proxy changes to a separate follow-on PR, as it seems to be needing more work.

**Comment by jrajahalme** [comment]
Split out the DNS proxy changes to a follow-on PR so that the egress Envoy L7 support can be shipped.

**Comment by jrajahalme** [comment]
test-me-please

**Pull request merged**